### PR TITLE
Parameterize block generation time related values

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

Three init flags are introduced
- `BlockGenerationIntervalFlag` determines block generation interval between two successive blocks (Only for CNs)
- `BlockGenerationTimeLimitFlag` sets the vm execution time limit during block generation (Only for CNs)
- `OpcodeComputationCostLimitFlag`  sets the computation cost limit for a tx

These are experimental flags for special usages. Some of them will be changed to governance items later.

**Block generation test result** 
image
- options: `--block-generation-interval 2 --block-generation-time-limit 800ms --opcode-computation-cost-limit 200000000`
- CN HW spec: m5.xlarge * 4
- Test case: cpuHeavyTx

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: work/worker.go [hunk 1]
```diff
@@ -485,24 +485,22 @@ func (self *worker) commitNewWork() {
 	self.currentMu.Lock()
 	defer self.currentMu.Unlock()
 
-	tstart := time.Now()
 	parent := self.chain.CurrentBlock()
 
 	// TODO-Klaytn drop or missing tx
+	tstart := time.Now()
```
Comment (yoomee1313) [inline work/worker.go:491]: Can I ask why you move `tstart := time.Now()`?

Comment (aidan-kwon) [reply work/worker.go:491]: @yoomee1313 I just want to assign and re-assign `tstart` and `tstamp` in sequencely. It's not a big deal.

Comment (yoomee1313) [reply work/worker.go:491]: I understood.. :)

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -993,6 +993,23 @@ var (
 		Name:  "config",
 		Usage: "TOML configuration file",
 	}
+	BlockGenerationIntervalFlag = cli.Int64Flag{
+		Name: "block-generation-interval",
```
Comment (jeongkyun-oh) [inline cmd/utils/flags.go:997]: How about using a dot `.` for separation for consistency with other flags?

File: params/protocol_params.go [hunk 3]
```diff
@@ -153,19 +153,33 @@ const (
 	TxDataGas uint64 = 100
 )
 
+const (
+	DefaultBlockGenerationInterval    = int64(1) // digit: seconds
```
Comment (jeongkyun-oh) [inline params/protocol_params.go:157]: `digit` -> `unit`?

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -993,6 +993,23 @@ var (
...
+		Usage: "(experimental option) Set the block generation interval in seconds. " +
+			"It should be equal or larger than 1",
```
Comment (winnie-byun) [inline cmd/utils/flags.go:999]:
```suggestion
			"It should be equal or larger than 1. This flag is only applicable to CN. ",
```

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -993,6 +993,23 @@ var (
...
+		Value: params.DefaultBlockGenerationInterval,
+	}
+	BlockGenerationTimeLimitFlag = cli.DurationFlag{
+		Name: "block-generation-time-limit",
+		Usage: "(experimental option) Set the vm execution time limit during block generation. " +
+			"Less than half of the block generation interval is recommended for this value",
```
Comment (winnie-byun) [inline cmd/utils/flags.go:1005]:
```suggestion
			"Less than half of the block generation interval is recommended for this value. " +
			"This flag is only applicable to CN.",
```

File: cmd/utils/flags.go [hunk 6]
```diff
@@ -993,6 +993,23 @@ var (
...
+		Value: params.DefaultBlockGenerationTimeLimit,
+	}
+	OpcodeComputationCostLimitFlag = cli.Uint64Flag{
+		Name:  "opcode-computation-cost-limit",
+		Usage: "(experimental option) Set the computation cost limit for a tx",
```
Comment (winnie-byun) [inline cmd/utils/flags.go:1010]:
```suggestion
		Usage: "(experimental option) Set the computation cost limit for a tx." +
		"Should set the same value within the network.",
```

File: cmd/utils/flags.go [hunk 7]
```diff
@@ -1580,6 +1597,19 @@ func SetKlayConfig(ctx *cli.Context, stack *node.Node, cfg *cn.Config) {
 		cfg.RPCGasCap = new(big.Int).SetUint64(ctx.GlobalUint64(RPCGlobalGasCap.Name))
 	}
 
+	// Only CNs could set BlockGenerationIntervalFlag and BlockGenerationTimeLimitFlag
+	if ctx.GlobalIsSet(BlockGenerationIntervalFlag.Name) {
+		params.BlockGenerationInterval = ctx.GlobalInt64(BlockGenerationIntervalFlag.Name)
```
Comment (winnie-byun) [inline cmd/utils/flags.go:1602]:
```suggestion
		params.BlockGenerationInterval = ctx.GlobalInt64(BlockGenerationIntervalFlag.Name)
		params.BlockGenerationTimeLimit = params.BlockGenerationInterval / 4
```
How about setting `BlockGenerationTimeLimit` to 1/4 of `BlockGenerationIntervalFlag` after?

**Review by winnie-byun — COMMENTED** [review]
LGTM except for minor comments.
Nice work :)

File: cmd/utils/flags.go (hunk 2 above)
Comment (aidan-kwon) [reply cmd/utils/flags.go:997]: @jeongkyun-oh A dot is used to group flags, but I don't think `block.generation.interval` is a propper hierarchical group.

File: cmd/utils/flags.go (hunk 7 above)
Comment (aidan-kwon) [reply cmd/utils/flags.go:1604]: @winnie-byun In that case, `BlockGenerationInterval` could be changed unintentionally. I think providing flexibility is better for this kind of experimental feature. Indeed, we don't have a rationale behind `BlockGenerationTimeLimit = BlockGenerationInterval / 4`

File: params/protocol_params.go (hunk 3 above)
Comment (aidan-kwon) [reply params/protocol_params.go:157]:
```suggestion
	DefaultBlockGenerationInterval    = int64(1) // unit: seconds
```

File: cmd/utils/flags.go (hunk 2 above)
Comment (jeongkyun-oh) [reply cmd/utils/flags.go:997]: Alright. Thanks.

**Review by winnie-byun — APPROVED** [review]

**Comment by winnie-byun** [comment]
@aidan-kwon Do you think we should consider putting these values at `ChainConfig`?

**Review by yoomee1313 — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]

**Comment by ehnuje** [comment]
> @aidan-kwon Do you think we should consider putting these values at `ChainConfig`?

@winnie-byun As it is test-purpose configurations, I think it is better to leave them as they are for now.

**Comment by aidan-kwon** [comment]
> @aidan-kwon Do you think we should consider putting these values at `ChainConfig`?

@winnie-byun As @ehnuje remarked, I think it is enough for experimental features. When these features are officially introduced, I will consider your suggestion also. Thank you.

**Pull request merged**